src/core/ext/filters/client_channel/resolver/binder/binder_resolver.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/string_util.h>
+
+#include "src/core/ext/filters/client_channel/resolver_registry.h"
+#include "src/core/ext/filters/client_channel/server_address.h"
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gpr/string.h"
+#include "src/core/lib/iomgr/resolve_address.h"
+#include "src/core/lib/iomgr/unix_sockets_posix.h"
+
+namespace grpc_core {
+namespace {
+
+class BinderResolver : public Resolver {
+ public:
+  BinderResolver(ServerAddressList addresses, ResolverArgs args)
+      : result_handler_(std::move(args.result_handler)),
+        addresses_(std::move(addresses)),
+        channel_args_(grpc_channel_args_copy(args.args)) {}
+
+  ~BinderResolver() override { grpc_channel_args_destroy(channel_args_); };
+
+  void StartLocked() override {
+    Result result;
+    result.addresses = std::move(addresses_);
+    result.args = channel_args_;
+    channel_args_ = nullptr;
+    result_handler_->ReturnResult(std::move(result));
+  }
+
+  void ShutdownLocked() override {}
+
+ private:
+  std::unique_ptr<ResultHandler> result_handler_;
+  ServerAddressList addresses_;
+  const grpc_channel_args* channel_args_ = nullptr;
+};
+
+grpc_error_handle BinderAddrPopulate(absl::string_view authority,
+                                     grpc_resolved_address* resolved_addr) {
+  if (std::string(authority).size() + 1 > sizeof(resolved_addr->addr)) {
+    return GRPC_ERROR_CREATE_FROM_CPP_STRING(std::string(authority) +
+                                             " is too long to be handled");
+  }
+  if (std::find_if(authority.begin(), authority.end(),
+                   [](char c) { return !isalnum(c); }) != authority.end()) {
+    return GRPC_ERROR_CREATE_FROM_CPP_STRING(std::string(authority) +
+                                             " contains invalid character");
+  }
+  strcpy(resolved_addr->addr, std::string(authority).c_str());
```

Comment:
It looks like this is just setting the `grpc_resolved_addr` (which is essentially a `struct sockaddr`) to the binder ID string.  This seems a little dangerous, since you're not setting an address type anywhere.  The `sockaddr` address type field will be garbage by anything that tries to interpret it that way.

The PR description says that the SubchannelConnector will know how to deal with this.  Where is that code?  I'd like to make sure that we have a good story for how this all fits together.

Reply:
It is actually a `char[]` (and it is `reinterpret_cast`ed to `struct sockaddr` when required)

```
struct grpc_resolved_address {
  char addr[GRPC_MAX_SOCKADDR_SIZE];
  socklen_t len;
};
```

In binder transport's case we only need a simple string so I directly store a c-style string in it. When parsing `resolved_addr->addr` (which is null-terminated) we can make sure that we don't accidentally read more than `sizeof(resolved_addr->addr)` bytes, so I think it is not that bad

> The PR description says that the SubchannelConnector will know how to deal with this. Where is that code? I'd like to make sure that we have a good story for how this all fits together.

Let me clean up the code and link it here later

Reply:
1. When user calls `CreateBinderChannel`, we notify Java to start trying to connect to the remote component (identifier is passed to Java)
https://github.com/sifmelcara/grpc/blob/21dfd08ef5329c98614e96fd04b38a4624271e9c/src/core/ext/transport/binder/client/channel_create.cc#L89-L93

Then we save the identifier in `GRPC_ARG_SERVER_URI` channel arg.

2. Java will notify us when the connection is established. We then save the mapping between identifier and endpoint binder in `EndpointBinderPool`.  https://github.com/sifmelcara/grpc/blob/21dfd08ef5329c98614e96fd04b38a4624271e9c/src/core/ext/transport/binder/client/endpoint_binder_pool.cc

3. When `SubchannelConnector::Connect` is called, we get the identifier from `args.address->addr` and use it to obtain the endpoint binder from `EndpointBinderPool` https://github.com/sifmelcara/grpc/blob/21dfd08ef5329c98614e96fd04b38a4624271e9c/src/core/ext/transport/binder/client/binder_connector.cc#L37-L57